x/dsmr/storage_test.go

```diff
@@ -369,3 +373,129 @@ func TestRestartSavedChunks(t *testing.T) {
 	storage = restart()
 	confirmChunkStorage(storage)
 }
+
+func TestChunkProducerRateLimiting(t *testing.T) {
+	chunk, err := newChunk(
+		UnsignedChunk[dsmrtest.Tx]{
+			Producer:    testDefaultProducer,
+			Beneficiary: codec.Address{},
+			Expiry:      1,
+			Txs:         []dsmrtest.Tx{{ID: ids.GenerateTestID(), Expiry: 1_000_000}},
+		},
+		[48]byte{},
+		[96]byte{},
+	)
+	require.NoError(t, err)
+	chunkSize := uint64(len(chunk.bytes))
+
+	testCases := []struct {
+		name                 string
+		expiryTimes          []int64
+		newChunkExpiry       int64
+		weightLimit          uint64
+		minTime              int64
+		acceptedChunksExpiry []int64
+		expectedErr          error
+	}{
+		{
+			name:           "success - first",
+			expiryTimes:    []int64{},
+			newChunkExpiry: 50,
+			weightLimit:    chunkSize * 5,
+			expectedErr:    nil,
+		},
+		{
+			name:           "success - after",
+			expiryTimes:    []int64{1},
+			newChunkExpiry: 50,
+			weightLimit:    chunkSize * 5,
+			expectedErr:    nil,
+		},
+		{
+			name:           "success - before",
+			expiryTimes:    []int64{50},
+			newChunkExpiry: 1,
+			weightLimit:    chunkSize * 5,
+			expectedErr:    nil,
+		},
+		{
+			name:           "fail - localized saturated range",
+			expiryTimes:    []int64{0, 50, 99, 150, 500},
+			newChunkExpiry: 75,
+			weightLimit:    chunkSize * 2,
+			expectedErr:    ErrChunkRateLimitSurpassed,
+		},
+		{
+			name:           "fail - localized saturated range with multiple elements",
+			expiryTimes:    []int64{0, 100, 120, 150, 200},
+			newChunkExpiry: 130,
+			weightLimit:    chunkSize * 3,
+			expectedErr:    ErrChunkRateLimitSurpassed,
+		},
+		{
+			name:                 "success - accepted block clear previous limit",
+			expiryTimes:          []int64{0, 50, 100},
+			newChunkExpiry:       150,
+			weightLimit:          chunkSize * 3,
+			expectedErr:          nil,
+			minTime:              10,
+			acceptedChunksExpiry: []int64{50},
+		},
```

Comment:
Should accepting a block clear it from the limit within a window?

I think it should still count against the rate limit because it was consumed within that window.

Reply:
It's just a matter of framing, right ? Just think about this as the side of unaccepted chunks per producer.
Can you think of any clear advantage of maintaining a separate list of accepted-but-not-yet-expired chunks ?

Reply:
Yup, I think it's preferable that accepting a chunk does not allow the same validator to consume the resources within that window of time, but this can be changed without a network upgrade, so think this is fine for now.

If we do this over a time window (irrespective of accepted/pending), then we have a clear upper bound on the max amount of chunk storage consumed by a single validator. If we clear the chunks when they are accepted, then we don't have a well defined limit because once the chunks are accepted the validator that produced it can consume more resources.